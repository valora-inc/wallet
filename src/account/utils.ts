import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
