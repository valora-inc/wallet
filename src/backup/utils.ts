import {
  CELO_DERIVATION_PATH_BASE,
  generateKeys,
  MnemonicLanguages,
} from '@celo/cryptographic-utils'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import * as bip39 from 'react-native-bip39'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { useDispatch, useSelector } from 'src/redux/hooks'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { ETHEREUM_DERIVATION_PATH } from 'src/web3/consts'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

const MNEMONIC_STORAGE_KEY = 'mnemonic'

export async function generateKeysFromMnemonic(mnemonic: string) {
  const wordCount = countMnemonicWords(mnemonic)
  const derivationPath = wordCount === 24 ? CELO_DERIVATION_PATH_BASE : ETHEREUM_DERIVATION_PATH
  return generateKeys(mnemonic, undefined, undefined, undefined, bip39, derivationPath)
}

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Recovery Phrase',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function isValidBackupPhrase(phrase: string) {
  const allowedPhraseLengths = [12, 24]
  const phraseLength = countMnemonicWords(formatBackupPhraseOnEdit(phrase))
  return !!phrase && allowedPhraseLengths.includes(phraseLength)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
