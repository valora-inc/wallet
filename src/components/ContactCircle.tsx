import * as React from 'react'
import { Image, StyleSheet, Text, View, ViewStyle } from 'react-native'
import User from 'src/icons/User'
import { Recipient } from 'src/recipients/recipient'
import Colors from 'src/styles/colors'
import { typeScale } from 'src/styles/fonts'

interface Props {
  style?: ViewStyle
  size?: number
  recipient: Recipient
  backgroundColor?: Colors
  foregroundColor?: Colors
  borderColor?: Colors
  DefaultIcon?: React.ComponentType<{ foregroundColor?: Colors; backgroundColor?: Colors }>
}

const DEFAULT_ICON_SIZE = 40

const getAddressBackgroundColor = (address: string) =>
  `hsl(${parseInt(address.substring(0, 5), 16) % 360}, 53%, 93%)` as Colors
const getAddressForegroundColor = (address: string) =>
  `hsl(${parseInt(address.substring(0, 5), 16) % 360}, 67%, 24%)` as Colors
const getNameInitial = (name: string) => name.charAt(0).toLocaleUpperCase()

function ContactCircle({
  size: iconSize = DEFAULT_ICON_SIZE,
  recipient,
  style,
  backgroundColor,
  foregroundColor,
  borderColor,
  DefaultIcon = ({ foregroundColor }) => <User color={foregroundColor} />,
}: Props) {
  const address = recipient.address
  const iconBackgroundColor = backgroundColor ?? getAddressBackgroundColor(address || '0x0')

  const renderThumbnail = () => {
    if (recipient.thumbnailPath) {
      return (
        <Image
          source={{ uri: recipient.thumbnailPath }}
          style={[
            styles.image,
            { height: iconSize, width: iconSize, borderRadius: iconSize / 2.0 },
          ]}
          resizeMode={'cover'}
        />
      )
    }

    const fontColor = foregroundColor ?? getAddressForegroundColor(address || '0x0')
    if (recipient.name) {
      const initial = getNameInitial(recipient.name)
      return (
        <Text
          allowFontScaling={false}
          style={[
            typeScale.labelMedium,
            { fontSize: iconSize / 2.0, color: fontColor, lineHeight: iconSize / 1.5 },
          ]}
        >
          {initial.toLocaleUpperCase()}
        </Text>
      )
    }

    return <DefaultIcon foregroundColor={fontColor} backgroundColor={iconBackgroundColor} />
  }

  return (
    <View style={[styles.container, style]}>
      <View
        style={[
          styles.icon,
          {
            backgroundColor: iconBackgroundColor,
            height: iconSize,
            width: iconSize,
            borderRadius: iconSize / 2,
          },
          borderColor && {
            borderColor,
            borderWidth: 1,
          },
        ]}
      >
        {renderThumbnail()}
      </View>
    </View>
  )
}

export default ContactCircle

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  icon: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  image: {
    margin: 'auto',
    alignSelf: 'center',
  },
})
