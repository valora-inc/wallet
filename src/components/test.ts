function groupNumber(value: string) {
  return value.replace(/\B(?=(\d{3})+(?!\d))(?<!\.\d*)/g, 'group')
}

function actualTest(value: string, decimalSeparator: string, groupingSeparator: string) {
  return groupNumber(value).replaceAll('.', decimalSeparator).replaceAll('group', groupingSeparator)
}

console.log(actualTest('1.234.567,9999', ',', '.')) // 1,234,567.9999
