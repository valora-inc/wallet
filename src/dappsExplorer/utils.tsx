import { DappV2WithCategoryNames } from 'src/dapps/types'

/***
 * Calculates a score based on the search term and dapp
 * - +2 if the name matches the search term in a case sensitive way
 * - +1 if the name matches the search term in a case insensitive way
 * - +1 if the description matches the search term in a case insensitive way
 * - +1 if the category matches the search term in a case insensitive way
 * @param dapp: dapp with categoryNames
 * @param searchTerm: the user input search string
 * @returns {number} the score of the dapp 0 - 5
 */
export const calculateSearchScore = (dapp: DappV2WithCategoryNames, searchTerm: string) => {
  if (searchTerm === '') return 0
  const nameMatchingCaseScore = dapp.name.includes(searchTerm) ? 2 : 0
  const nameScore = dapp.name.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const descriptionScore = dapp.description.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const categoryScore = dapp.categoryNames.some((category) =>
    category.toLowerCase().includes(searchTerm.toLowerCase())
  )
    ? 1
    : 0
  return nameMatchingCaseScore + nameScore + descriptionScore + categoryScore
}
