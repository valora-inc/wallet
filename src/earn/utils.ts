import BigNumber from 'bignumber.js'
import { getFeatureGate } from 'src/statsig'
import { StatsigFeatureGates } from 'src/statsig/types'
import { SwapTransaction } from 'src/swap/types'
import { Network, NetworkId } from 'src/transactions/types'
import { INTERNAL_RPC_SUPPORTED_NETWORKS } from 'src/viem'
import { networkIdToNetwork } from 'src/web3/networkConfig'

/**
 * Helper function to check if gas is subsidized for the given networkId
 * @param networkId - networkId to check
 * @returns boolean
 */
export function isGasSubsidizedForNetwork(networkId: NetworkId) {
  return (
    (INTERNAL_RPC_SUPPORTED_NETWORKS as readonly Network[]).includes(
      networkIdToNetwork[networkId]
    ) && getFeatureGate(StatsigFeatureGates.SUBSIDIZE_STABLECOIN_EARN_GAS_FEES)
  )
}

/**
 * Gets the swap to amount in decimals given a swap transaction and the swap
 * from amount in decimals. This could also be taken from the buy amount, but
 * that is in wei and we would need the token decimals. This follows what is
 * done in the SwapScreen
 * @param object.swapTransaction - swap transaction object
 * @param object.fromAmount - swap from amount in decimals (usually from the user input)
 * @returns swap to amount in decimals
 */
export function getSwapToAmountInDecimals({
  swapTransaction,
  fromAmount,
}: {
  swapTransaction: SwapTransaction
  fromAmount: BigNumber
}) {
  return fromAmount.multipliedBy(swapTransaction.price)
}
