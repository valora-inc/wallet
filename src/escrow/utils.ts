import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToE164NumberSelector } from 'src/identity/selectors'
import { ContactRecipient } from 'src/recipients/recipient'
import { useSelector } from 'src/redux/hooks'

export const useEscrowPaymentRecipient = (payment: EscrowedPayment): ContactRecipient => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToE164Number = useSelector(identifierToE164NumberSelector)
  const recipientCache = useSelector((state) => state.recipients.phoneRecipientCache)
  const phoneNumber = identifierToE164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber] ?? {
    name: phoneNumber,
    e164PhoneNumber: phoneNumber,
    contactId: '',
  }

  return recipient
}
