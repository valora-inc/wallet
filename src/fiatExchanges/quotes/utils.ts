import { SettlementEstimation, SettlementTime } from 'src/fiatExchanges/quotes/constants'
import Logger from 'src/utils/Logger'

const TAG = 'fiatExchanges/quotes/utils'

export const getSettlementTimeString = (
  settlementEstimation: SettlementEstimation,
  settlementTimeStrings: Record<SettlementTime, string>
) => {
  switch (settlementEstimation.settlementTime) {
    case SettlementTime.LESS_THAN_ONE_HOUR:
      return { timeString: settlementTimeStrings[SettlementTime.LESS_THAN_ONE_HOUR] }
    case SettlementTime.LESS_THAN_X_HOURS:
    case SettlementTime.LESS_THAN_X_DAYS:
      return {
        timeString: settlementTimeStrings[settlementEstimation.settlementTime],
        upperBound: settlementEstimation.upperBound,
      }
    case SettlementTime.X_TO_Y_HOURS:
    case SettlementTime.X_TO_Y_DAYS:
      return {
        timeString: settlementTimeStrings[settlementEstimation.settlementTime],
        lowerBound: settlementEstimation.lowerBound,
        upperBound: settlementEstimation.upperBound,
      }
    default:
      // Should not be reachable
      Logger.warn(TAG, 'Unexpected settlement time value in settlement estimation')
      return {}
  }
}
