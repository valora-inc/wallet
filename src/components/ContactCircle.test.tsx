import { render } from '@testing-library/react-native'
import * as React from 'react'
import { Image } from 'react-native'
import { Provider } from 'react-redux'
import ContactCircle from 'src/components/ContactCircle'
import { RecipientType } from 'src/recipients/recipient'
import { createMockStore } from 'test/utils'

const mockAddress = '0x123456'
const mockName = 'Mock name'

const mockStore = createMockStore()

describe('ContactCircle', () => {
  describe('when given recipient with only address', () => {
    it('uses DefaultAvatar svg', () => {
      const wrapper = render(
        <Provider store={mockStore}>
          <ContactCircle
            size={30}
            recipient={{
              address: mockAddress,
              recipientType: RecipientType.Address,
            }}
          />
        </Provider>
      )
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when has a thumbnail and name', () => {
    it('renders image', () => {
      const mockThumbnnailPath = './test.jpg'
      const { UNSAFE_getByType } = render(
        <Provider store={mockStore}>
          <ContactCircle
            size={30}
            recipient={{
              name: mockName,
              address: mockAddress,
              thumbnailPath: mockThumbnnailPath,
              recipientType: RecipientType.Address,
            }}
          />
        </Provider>
      )
      expect(UNSAFE_getByType(Image).props.source).toEqual({ uri: './test.jpg' })
    })
  })
  describe('when has a name but no picture', () => {
    it('renders initial', () => {
      const { getByText } = render(
        <Provider store={mockStore}>
          <ContactCircle
            size={30}
            recipient={{
              name: mockName,
              address: mockAddress,
              recipientType: RecipientType.Address,
            }}
          />
        </Provider>
      )
      expect(getByText(mockName[0])).toBeTruthy()
    })
  })
})
