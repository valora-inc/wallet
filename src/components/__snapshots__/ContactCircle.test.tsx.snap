// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactCircle when given recipient with only address uses DefaultAvatar svg 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "hsl(291, 53%, 93%)",
          "borderRadius": 15,
          "height": 30,
          "width": 30,
        },
      ]
    }
  >
    <svg
      fill="hsl(291, 53%, 93%)"
      height="40"
      style={Object {}}
      viewBox="0 0 40 40"
      width="40"
    />
  </View>
</View>
`;
